frontend/dds_mvp.js

```diff
@@ -54,90 +69,124 @@ function fillFormWithPartScoreTestData() {
+        hands.unshift(west);
+    }
 
-        for (var suit_index = 0; suit_index < 4; suit_index++) {
-            var suit = SUITS[suit_index];
-            var element_index = new_direction + "_" + suit;
-            var new_element = document.getElementById(element_index);
-            new_element.value = old_hands[direction_index][suit_index];
-        }
+    for (const element of hand_elements()) {
+        element.value = hands.shift();
     }
 }
 
 function collectHands() {
     var hands = {};
 
-    for (var direction of DIRECTIONS) {
-        var direction_letter = direction.charAt(0).toUpperCase();
-        hands[direction_letter] = [];
+    for (const ds of directions_and_suits()) {
+        var direction_letter = ds.direction.charAt(0).toUpperCase();
 
-        for (var suit of SUITS) {
-            var suit_letter = suit.charAt(0).toUpperCase();
-            var element_index = direction + "_" + suit;
-            var holding = document.getElementById(element_index).value;
+        hands[direction_letter] = hands[direction_letter] || [];
 
-            for (var card of holding) {
-                hands[direction_letter].push(suit_letter + card.toUpperCase());
-            }
-       }
+        var suit_letter = ds.suit.charAt(0).toUpperCase();
+        var element_index = ds.direction + "_" + ds.suit;
+        var holding = document.getElementById(element_index).value;
+
+        for (const card of holding) {
+            hands[direction_letter].push(suit_letter + card.toUpperCase());
+        }
     }
 
     return hands;
 }
 
-function sendJSON(){
-    var xhr = new XMLHttpRequest();
-    const URL = "http://localhost:5000/api/dds-table/";
-
-    // This fails as of 2020-06-21 due to:
-    // Access from origin 'null' has been blocked by CORS policy
-    // const URL = "https://dds.hackathon.globalbridge.app/api/dds-table/";
+function inputIsValid(hands) {
+    // TODO Make sure no card is repeated.
 
-    xhr.open("POST", URL, true);
+    for (const direction in hands) {
+        const hand = hands[direction];
 
-    xhr.setRequestHeader("Content-Type", "application/json");
+        if (hand.length != 13) {
+            return "Please enter 13 cards per hand.";
+        }
 
-    xhr.onreadystatechange = function () {
-        if (xhr.readyState === 4 && xhr.status === 200) {
-            document.getElementById("result").innerHTML = this.responseText;
+        for (const card of hand) {
+            const pip = card.substring(1);
+            if (!PIPS.includes(pip)) {
+                return "Please use only these pips: " + PIPS;
+            }
         }
-    };
+    }
+
+    return "";
+}
+
+function pageLoad() {
+    document.getElementById("valid-pips").innerHTML = PIPS;
+}
 
+function sendJSON() {
     var hands = collectHands();
-    var deal = { "hands": hands };
-    var data = JSON.stringify(deal);
-    xhr.send(data);
+
+    const error_message = inputIsValid(hands);
+
+    if (error_message.length == 0) {
```

Comment:
A good practice would be having error check only having failure branch which returns. Then rest of function body would avoid increased indention. In this case something like:

if (error_messa.length != 0) {
document.getElementById("result").innerHTML = error_message;
return;
}

Reply:
Will do.

Reply:
Done and pushed. Please have another look.